### DISCUSSION TITLE
Should client.multicall validate that functionNames are in the ABI?

### DISCUSSION THREAD

**Discussion by hammeiam** [wevm/viem#377] [Question] [answered]
I really love how contract instances and things like `getAbiItem` will both suggest names from the ABI and complain when you don't use a proper name. I was wondering if the multicall function should do the same. Thanks!

**Comment by tmm** [comment]
It should already do that! Have an example? Works fine in this [TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQBQK4BGANsMgMIloB2MANHABYwxhwC+cAZlBCHAOQA3YGhD8AUKEixEcEBmBUqaeB268Bw0QHpkDBVQDOEqdHhIA7hgDmIYAFlFMAKIAPDOCJoAggWDsuHj5+DD8YAE8wNG0AE2BDGG1sBIlxZAgjeGRKGjgAXhR0LFxCEnIcmAAKBHE4FH1FAC45A2U6WrgYKAwjaRhmphZKgEpacTZhtIqAOhA8IhhSDCIiao70mm7kGENmgG0Oupq6k7gMGJj0Q12BAAZXQDICfjHTutDgZqtbByc3DzAvL5gC9Xpw8FRtsAMgA5DxoZr8QzhEAECBEZ6HE7abRwAB6AH5MWwQUdMW8Llcbvx7k8SSd3p8bHZHDQ-p4fH46XUwRDFjC4QiCMsesg0AB5TgY15wbF4wnSjBQaw3PbUx78AC6dOJmOOr3OlzQ1wRNKl+r8jO+LJc7nZQK5XHBkP5IHhAh5zqosNdABEIEboRAbfEYGa6jq6lrxpMgA).

**Comment by hammeiam** [comment]
I sure do! One image is of `createEventFilter` correctly alerting me to an invalid event name. The other is of `multicall` allowing me to pass a clearly invalid function name.

<img width="662" alt="Screen Shot 2023-04-14 at 9 46 59 PM" src="https://user-images.githubusercontent.com/7607886/232179455-6ae6a85a-cf8a-4201-bd20-a56caa1292f1.png">
<img width="510" alt="Screen Shot 2023-04-14 at 9 46 31 PM" src="https://user-images.githubusercontent.com/7607886/232179468-3cdde77e-23cd-4556-9bb1-aa8cb7823a5c.png">

**Reply by tmm** [reply, marked as answer]
This is a limitation of TypeScript at the moment and how it handles higher-order functions. See https://github.com/microsoft/TypeScript/issues/29841 Workaround is to cast each mapped value using `ContractFunctionConfig`:

```ts
import type { ContractFunctionConfig } from 'viem'

const res = await publicClient.multicall({
  allowFailure: false,
  contracts: tokenAddresses.map(
    (x) =>
      ({
        address: x,
        abi: erc20Abi,
        functionName: 'balanceOf',
        args: ['0x'],
      } as ContractFunctionConfig<
        typeof erc20Abi,
        'balanceOf',
        'pure' | 'view'
      >),
  ),
})
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQBQK4BGANsMgMIloB2MANHABYwxj1kQ1QbIwBieVHsA7sqAM2ABzOAF84YqBBBwA5ADdgaECoBQoSLERwQGYFSpp4chUtUatAemQNTVAM6790eEjRRkAEwADACCBMCy8orKKhjhMACeYGgOACbAbjAO2Jm6OsgcmXBghCTklDRwALwo6Fi4paQUmjQAFAg6cCguZgBcxq4WdJ1wMFzuBjD9TCytAJS0OjJzOvmF8DAQANbUIamp6G5uaG7VcADaKkEAHir0V7cAunAYpwXuMGsfcIdnGADupngJWITQqMAAdCA8EQYKQMEQiO0RgiiBB-rxTEQ8Oh+mIEcdFl13mNuDA3P1NjsqHsDidjm4oRgwK0Rl1Wtc5tUAHxsrpwZH8oUvfaHClwa5E4VdOLAfp+QKhcJS6ViARCDgAOQwIDQ-RUBARGEEaAA8mI7nyhRgoJJxZcbipHir+XJXnBRKSePxBHCRBwJJIADxW-mJZIQMRwBXBMLAF1Cg1Gk3my3SroqEroFRwAA+dk0-100u5CxGZeWOkOOgcDjgAD0APw6IA)
